addons/product/models/product_product.py

```diff
@@ -354,10 +354,15 @@ def _compute_partner_ref(self):
 
     def _compute_product_document_count(self):
         for product in self:
-            product.product_document_count = product.env['product.document'].search_count([
+            product_doc_count = product.env['product.document'].search_count([
                 ('res_model', '=', 'product.product'),
                 ('res_id', 'in', product.ids),
             ])
+            template_doc_count = product.env['product.document'].search_count([
+                ('res_model', '=', 'product.template'),
+                ('res_id', 'in', product.product_tmpl_id.ids),
+            ])
+            product.product_document_count = product_doc_count + template_doc_count
```

Comment:
Please combine the two domains (and two `search_count` calls) with an `Domain.OR` to avoid triggering two sql queries and improve performance ;)